### PR TITLE
Encourage users to explicitly pass resources when logging langchain models with dependencies

### PR DESCRIPTION
**Pull request by smurching**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/smurching/mlflow/pull/13708?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13708/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13708
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

See title - encourage users to explicitly pass resources when logging langchain models with dependencies, since it's more robust than dependency auto-detection

### How is this PR tested?
* Updated unit tests
* Checked docs preview and made sure the updated docstring renders well:
image

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import os
 import tempfile
 import warnings
-from typing import Any, Iterator, Optional, Union
+from typing import Any, Iterator, Optional, Union, List
```
Comment (smurching) [inline mlflow/langchain/__init__.py:24]: /autoformat

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -364,6 +364,13 @@ def load_retriever(persist_directory):
     needs_databricks_auth = False
     if Version(langchain.__version__) >= Version("0.0.311") and mlflow_model.resources is None:
         if databricks_resources := _detect_databricks_dependencies(lc_model):
+            logger.info("Attempting to auto-detect Databricks resource dependencies for the "
```
[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/langchain/__init__.py:367]: Good question - this is actually printed after we detect dependencies, but I didn't include which ones were autoinferred, let me add that in

[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/langchain/__init__.py:367]: Actually, I'm a bit worried the list of dependencies may be large & bloat logs e.g. if the agent has a UC schema of function dependencies, let's keep the log message simple for now. I think if needed, we should add some docs + APIs for better debugging/inspecting model resource dependencies, so customers can do the comparison you mentioned easily

[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/langchain/__init__.py:367]: e.g. document how to load the model back and inspect its resources using `mlflow.models.Model(...).resources`

**Comment by BenWilson2** [comment]
Can we add a note about this here? https://mlflow.org/docs/latest/llms/langchain/index.html#dependency-management I think this is pretty important and warrants special mention within the LangChain docs.

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -364,6 +364,13 @@ def load_retriever(persist_directory):
...
+                        "current langchain model. Dependency auto-detection is "
+                        "best-effort and may not capture all dependencies of your langchain "
+                        "model, resulting in authorization errors when serving or querying your model. "
+                        "We recommend that you explicitly pass `resources` "
+                        "to mlflow.langchain.log_model() to ensure authorization to dependent resources "
+                        "succeeds when the model is deployed.")
```
Comment (BenWilson2) [inline mlflow/langchain/__init__.py:373]:
```suggestion
                        "model, resulting in authorization errors when serving or querying "
                        "your model. We recommend that you explicitly pass `resources` "
                        "to mlflow.langchain.log_model() to ensure authorization to "
                        "dependent resources succeeds when the model is deployed.")
```

**Comment by smurching** [comment]
> Can we add a note about this here? https://mlflow.org/docs/latest/llms/langchain/index.html#dependency-management I think this is pretty important and warrants special mention within the LangChain docs.

@BenWilson2 curious what you'd suggest mentioning here, this section and [the page it links to](https://mlflow.org/docs/latest/model/dependencies.html) seem more about managing python environment dependencies, which does work automatically for LangChain models. For now I'll just clarify to update that "code and environment dependencies" are the ones that are inferred automatically

The resource dependency auto-inference behavior here is specific to Databricks and [is documented here](https://docs.databricks.com/en/generative-ai/agent-framework/deploy-agent.html#authentication-for-dependent-resources), not sure it makes sense to mention in open-source docs

File: docs/source/llms/langchain/index.rst [hunk 4]
```diff
@@ -25,7 +25,12 @@ LangChain's flexibility in experimenting with various agents, tools, and retriev
 
 Dependency Management
 ^^^^^^^^^^^^^^^^^^^^^
-Deploy your LangChain application with confidence, leveraging MLflow's ability to `manage and record all external dependencies <../../model/dependencies.html>`_ automatically. This ensures consistency between development and production environments, reducing deployment risks with less manual intervention.
+Deploy your LangChain application with confidence, leveraging MLflow's ability to `manage and record code and environment dependencies <../../model/dependencies.html>`_ automatically.
+You can also explicitly declare external resource dependencies, like the LLM serving endpoint or vector search index queried by your LangChain application.
+These dependencies are tracked by MLflow as model metadata, so that downstream serving systems can ensure authentication from your
+deployed LangChain application to these dependent resources just works.
+
+These features ensures consistency between development and production environments, reducing deployment risks with less manual intervention.
```
Comment (BenWilson2) [inline docs/source/llms/langchain/index.rst:33]:
```suggestion
These features ensure consistency between development and production environments, reducing deployment risks with less manual intervention.
```

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**